### PR TITLE
Improve X509Chain handling of NotSignatureValid on Linux

### PR DESCRIPTION
**Pull request by bartonjs**
Addresses #35654 in master.

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs [hunk 1]
```diff
@@ -876,8 +900,26 @@ private static void AddToStackAndUpRef(IntPtr cert, SafeX509StackHandle stack)
...
 
+            internal bool AbortOnSignatureError { get; }
+            internal bool AbortedForSignatureError { get; private set; }
+
+            internal WorkingChain()
+            {
+                AbortOnSignatureError = s_defaultAbort;
+            }
+
+            internal WorkingChain(bool abortOnSignatureError)
```
Comment (GrabYourPitchforks) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs]: Nit: Make this ctor private and have the parameterless ctor call it.

File: src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c [hunk 2]
```diff
@@ -682,6 +688,126 @@ int32_t CryptoNative_X509StoreCtxCommitToChain(X509_STORE_CTX* storeCtx)
...
+        X509* last = sk_X509_value(chain, errorDepth);
+
+        // If the last element is in the trust store we need to build a new trust store.
+        if (X509_STORE_CTX_get1_issuer(&root, storeCtx, last))
+        {
+            if (root == last)
+            {
+                // We know it's a non-zero refcount after this because last has one, too.
+                // So go ahead and undo the get1.
+                X509_free(root);
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c]: This looks strange.  We're freeing something that's part of / still referenced by the store?

File: src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c [hunk 3]
```diff
@@ -682,6 +688,126 @@ int32_t CryptoNative_X509StoreCtxCommitToChain(X509_STORE_CTX* storeCtx)
...
+                store = tmpNew;
+                chainLength--;
+            }
+            else
+            {
+                // This really shouldn't happen, since if we could have resolved it now
+                // it should have resolved during chain walk.
+                //
+                // But better safe than sorry.
+                X509_free(root);
```
Comment (stephentoub) [inline src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c]: If we do the X509_free(root) as the first thing in both the if and the else, couldn't we just move it to before the if and eliminate the else entirely?

File: src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c (hunk 2 above)
Comment (bartonjs) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c]: > Or the get1_issuer call incremented the ref count and we're just undoing that?

Yeah, that.  `get0` means "I'm giving it to you without up_ref", `get1` means "I'm giving it to you, and it was up_ref'd"

File: src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c (hunk 3 above)
Comment (bartonjs) [reply src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c]: We.... could.  But I was a bit hesitant to let a flow exist where we downref root to 0 (when it doesn't match) but we end up talking about it still.

**Pull request merged**